balancer/pickfirst_leaf/pickfirst_leaf.go

```diff
@@ -0,0 +1,627 @@
+	if b.logger.V(2) {
+		b.logger.Infof("Received error from the name resolver: %v", err)
+	}
+	if b.state == connectivity.Shutdown {
+		return
+	}
+	// The picker will not change since the balancer does not currently
+	// report an error.
+	if b.state != connectivity.TransientFailure {
+		return
+	}
+
+	for _, sd := range b.subConns.Values() {
+		sd.(*scData).subConn.Shutdown()
+	}
+	for _, k := range b.subConns.Keys() {
+		b.subConns.Delete(k)
+	}
+	b.addressIndex.updateEndpointList(nil)
+	b.state = connectivity.TransientFailure
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: connectivity.TransientFailure,
+		Picker:            &picker{err: fmt.Errorf("name resolver error: %v", err)},
+	})
+}
+
+func (b *pickfirstBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	errCh := make(chan error, 1)
+	b.serializer.ScheduleOr(func(_ context.Context) {
+		err := b.updateClientConnState(state)
+		errCh <- err
+	}, func() {
+		close(errCh)
+	})
+	return <-errCh
+}
+
+func (b *pickfirstBalancer) updateClientConnState(state balancer.ClientConnState) error {
+	if b.state == connectivity.Shutdown {
+		return fmt.Errorf("balancer is already closed")
+	}
+	if len(state.ResolverState.Addresses) == 0 && len(state.ResolverState.Endpoints) == 0 {
+		// Cleanup state pertaining to the previous resolver state.
+		// Treat an empty address list like an error by calling b.ResolverError.
+		b.state = connectivity.TransientFailure
+		b.resolverError(errors.New("produced zero addresses"))
+		return balancer.ErrBadResolverState
+	}
+	// We don't have to guard this block with the env var because ParseConfig
+	// already does so.
+	cfg, ok := state.BalancerConfig.(pfConfig)
+	if state.BalancerConfig != nil && !ok {
+		return fmt.Errorf("pickfirst: received illegal BalancerConfig (type %T): %v", state.BalancerConfig, state.BalancerConfig)
+	}
+
+	if b.logger.V(2) {
+		b.logger.Infof("Received new config %s, resolver state %s", pretty.ToJSON(cfg), pretty.ToJSON(state.ResolverState))
+	}
+
+	newEndpoints := state.ResolverState.Endpoints
+	if len(newEndpoints) == 0 {
+		newEndpoints = make([]resolver.Endpoint, len(state.ResolverState.Addresses))
+		// Convert addresses to endpoints.
+		for i, a := range state.ResolverState.Addresses {
+			newEndpoints[i].Attributes = a.BalancerAttributes
+			newEndpoints[i].Addresses = []resolver.Address{a}
+		}
+	}
+
+	// Since we have a new set of addresses, we are again at first pass
+	b.firstPass = true
+	newEndpoints = deDupAddresses(newEndpoints)
+
+	// Perform the optional shuffling described in gRFC A62. The shuffling will
+	// change the order of endpoints but not touch the order of the addresses
+	// within each endpoint. - A61
+	if cfg.ShuffleAddressList {
+		newEndpoints = append([]resolver.Endpoint{}, newEndpoints...)
+		internal.ShuffleAddressListForTesting.(func(int, func(int, int)))(len(newEndpoints), func(i, j int) { newEndpoints[i], newEndpoints[j] = newEndpoints[j], newEndpoints[i] })
```

Comment:
The problem that I'm trying to point out is that:
- `pickfirst` sets `ShuffleAddressListForTesting` to use `rand.Shuffle` in its `init` function
- `pickfirst` also directly uses `ShuffleAddressListForTesting` in `UpdateClientConnState` to get the actual shuffle function

So, for someone reading the code, it can be a little misleading as to why `pickfirst` is getting the shuffle func from a var with a `ForTesting` suffix. 

Instead what I'm suggesting is to do something like:
```golang
package pickfirst

var shuffleFunc func(n int, swap func(i, j int))

func init() {
  shuffleFunc = rand.Shuffle
  internal.ShuffleAddressListForTesting	= func(sf func(n int, swap func(i, j int))) {
    shuffleFunc = sf
  }  
}

func (...) UpdateClientConnState(...) {
  // Directly use `shuffleFunc` to do the shuffling
}
```

Note that I've changed the signature of `internal.ShuffleAddressListForTesting` to accept a `func`. We could rename it to something like `internal.PickfirstShuffleFunc`. Also, note that we could completely move this var to the `pickfirst` package and have our tests directly use that, but would pollute the external API.

Again, as I mentioned previously, there is absolutely no need to do this is in this PR. We should do it as a follow-up. Thanks.

Reply:
Understood. I've opened https://github.com/grpc/grpc-go/issues/7515 to track the cleanup. I will pick it up once this PR is merged and change both the old and new `pickfirsts`.

Reply:
Thanks :)